src/components/TodoContext.tsx

```diff
@@ -0,0 +1,99 @@
+import React, { useEffect, useState } from 'react';
+import { TodosContextType } from '../types/TodoContext';
+import { Todo } from '../types/Todo';
+import { createTodo, deleteTodo, getTodos } from '../api/todos';
+import { ErrorType } from '../types/Error';
+
+const USER_ID = 11121;
+
+export const TodosContext = React.createContext<TodosContextType | null>(null);
+
+type Props = {
+  children: React.ReactNode,
+};
+
+export const TodosProvider: React.FC<Props> = ({ children }) => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [error, setError] = useState('');
+  const [disabledInput, setDisabledInput] = useState(false);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [isLoading, setIsLoading] = useState(false);
+
+  useEffect(() => {
+    getTodos()
+      .then((todosFromServer) => {
+        setTodos(todosFromServer);
+      })
+      .catch(() => {
+        setError(ErrorType.fetchError);
+      });
+  }, []);
+
+  const resetError = () => {
+    setError('');
+  };
+
+  const handleSetError = (errorType: string) => {
+    setError(errorType);
+  };
+
+  const addTodo = (title: string) => {
+    const newTempTodo: Todo = {
+      id: 0,
+      userId: USER_ID,
+      title,
+      completed: false,
+    };
+
+    setTempTodo(newTempTodo);
+    setDisabledInput(true);
+
+    createTodo(newTempTodo)
+      .then((createdTodo) => {
+        setTempTodo(null);
+        setTodos(prevTodos => [createdTodo, ...prevTodos]);
+        setError('');
+      })
+      .catch(() => {
+        setTempTodo(null);
+        setError(ErrorType.addTodo);
+      })
+      .finally(() => {
+        setDisabledInput(false);
+      });
+  };
+
+  const removeTodo = (todoId: number) => {
+    setIsLoading(true);
+
+    deleteTodo(todoId)
+      .then(() => {
+        setTodos(
+          currentTodos => currentTodos.filter(todo => todo.id !== todoId),
+        );
+      })
+      .catch(() => {
+        setError(ErrorType.deleteTodo);
+      })
+      .finally(() => {
+        setIsLoading(false);
+      });
+  };
+
+  return (
+    <TodosContext.Provider value={{
+      todos,
+      error,
+      resetError,
+      addTodo,
+      removeTodo,
+      handleSetError,
+      disabledInput,
+      tempTodo,
+      isLoading,
+    }}
```

Comment:
The 'value' object passed as the value prop to the Context provider changes every render.
Might be better to consider wrapping it in a useMemo and moving setState functions to a different provider, they wouldn't change each render